src/dune_engine/build_system.ml

```diff
@@ -40,11 +40,14 @@ end = struct
     match Path.as_in_build_dir path with
     | Some path -> mkdir_p path
     | None -> (
-      Fs_memo.path_exists path >>| function
+      Fs_memo.dir_exists path >>| function
       | true -> ()
       | false ->
+        (* CR-someday amokhov: I think this case is impossible because we only
+           call [mkdir_p_or_assert_existence] for directories in [Chdir] actions
+           but those are checked to be in the build directory. *)
         User_error.raise ~loc
-          [ Pp.textf "%S does not exist" (Path.to_string_maybe_quoted path) ])
+          [ Pp.textf "Directory %S does not exist" (Path.to_string path) ])
```

Comment:
I tried to trigger this error in a test but couldn't. I guess we should just use `mkdir_p` instead of `mkdir_p_or_assert_existence` for simplicity.

Reply:
I kind of remember something about this, i.e. allowing such external chdirs for some reason. /cc @rgrinberg does that ring a bell?

Reply:
I found this: 

https://github.com/ocaml/dune/blob/c3d9b819dc3f19ef75fdcd59e1ccaa0394fc3064/src/dune_rules/action_unexpanded.ml#L12-L19

Which seems to indicate all paths of user-written actions must be in build directory. Perhaps, if we create one in a custom rule we can indeed violate this.

Reply:
Hmm, it's possible that it was due to a builtin rule. Did you `git blame` `mkdir_p_or_...` to see why this was added?

Reply:
I think this was added in this commit: https://github.com/ocaml/dune/commit/d5746c6b603f51c6035ea75d501b2ddbac6c4b54

Reply:
And here is the corresponding PR: https://github.com/ocaml/dune/pull/2165

Reply:
Alright, so looks like we were trying to be conservative, but one code path is/(has become?) unreachable. Might be worth making the "not a build dir" an error here and seeing if anything breaks.

Reply:
The testsuite still passes if I raise an error for a non-build dir, so maybe we should just do that. I'll raise a separate PR for this.